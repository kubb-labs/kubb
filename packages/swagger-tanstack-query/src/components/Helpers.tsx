import path from 'node:path'
import url from 'node:url'

import { File } from '@kubb/react'

import type { ReactNode } from 'react'

export const templatesPath = path.resolve(path.dirname(url.fileURLToPath(import.meta.url)), '../templates')

const typesPath = path.resolve(templatesPath, './types.ts')

export function Helpers(): ReactNode {
  return (
    <>
      <File.Source path={typesPath} print removeComments></File.Source>
    </>
  )
}
