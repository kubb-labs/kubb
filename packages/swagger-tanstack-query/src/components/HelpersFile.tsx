import path from 'node:path'
import url from 'node:url'

import { File } from '@kubb/react'

import type { ReactNode } from 'react'

export const templatesPath = path.resolve(path.dirname(url.fileURLToPath(import.meta.url)), '../templates')

const typesPath = path.resolve(templatesPath, './types.ts')

type Props = {
  id?: string
  path: string
}

export function HelpersFile({ id, path: resolvedPath }: Props): ReactNode {
  return (
    <File id={id} override baseName={'types.ts'} path={resolvedPath}>
      <File.Source path={typesPath} print removeComments />
    </File>
  )
}
