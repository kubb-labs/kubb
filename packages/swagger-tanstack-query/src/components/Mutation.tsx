import { PackageManager } from '@kubb/core'
import transformers from '@kubb/core/transformers'
import { FunctionParams, URLPath } from '@kubb/core/utils'
import { Editor, File, Function, usePlugin } from '@kubb/react'
import { useGetOperationFile, useOperation, useOperationName, useOperationSchemas } from '@kubb/swagger/hooks'
import { getASTParams, getComments, isRequired } from '@kubb/swagger/utils'
import { pluginKey as swaggerTsPluginKey } from '@kubb/swagger-ts'

import { getImportNames } from '../utils.ts'
import { MutationImports } from './MutationImports.tsx'
import { SchemaType } from './SchemaType.tsx'

import type { HttpMethod } from '@kubb/swagger/oas'
import type { ReactNode } from 'react'
import type { FileMeta, PluginOptions } from '../types.ts'

type TemplateProps = {
  /**
   * Name of the function
   */
  name: string
  /**
   * Parameters/options/props that need to be used
   */
  params: string
  mutateParams: string
  /**
   * Generics that needs to be added for TypeScript
   */
  generics?: string
  /**
   * ReturnType(see async for adding Promise type)
   */
  returnType?: string
  /**
   * Options for JSdocs
   */
  JSDoc?: {
    comments: string[]
  }
  hook: {
    name: string
    generics?: string
    children?: string
  }
  client: {
    method: HttpMethod
    generics: string
    withQueryParams: boolean
    withPathParams: boolean
    withData: boolean
    withHeaders: boolean
    path: URLPath
  }
  dataReturnType: NonNullable<PluginOptions['options']['dataReturnType']>
}

function Template({ name, generics, returnType, params, mutateParams, JSDoc, client, hook, dataReturnType }: TemplateProps): ReactNode {
  const isV5 = new PackageManager().isValidSync(/@tanstack/, '>=5')
  const clientOptions = [
    `method: "${client.method}"`,
    `url: ${client.path.template}`,
    client.withQueryParams ? 'params' : undefined,
    client.withData ? 'data' : undefined,
    client.withHeaders ? 'headers: { ...headers, ...clientOptions.headers }' : undefined,
    '...clientOptions',
  ].filter(Boolean)

  const resolvedClientOptions = `${transformers.createIndent(4)}${clientOptions.join(`,\n${transformers.createIndent(4)}`)}`

  if (isV5) {
    return (
      <Function export name={name} params={params} JSDoc={JSDoc}>
        {`
         const { mutation: mutationOptions, client: clientOptions = {} } = options ?? {}

         return ${hook.name}({
           mutationFn: async(${mutateParams}) => {
            ${hook.children || ''}
             const res = await client<${client.generics}>({
              ${resolvedClientOptions}
             })

             return ${dataReturnType === 'data' ? 'res.data' : 'res'}
           },
           ...mutationOptions
         })`}
      </Function>
    )
  }

  return (
    <Function export name={name} generics={generics} returnType={returnType} params={params} JSDoc={JSDoc}>
      {`
       const { mutation: mutationOptions, client: clientOptions = {} } = options ?? {}

       return ${hook.name}<${hook.generics}>({
         mutationFn: async(${mutateParams}) => {
          ${hook.children || ''}
           const res = await client<${client.generics}>({
            ${resolvedClientOptions}
           })

           return ${dataReturnType === 'data' ? 'res.data' : 'res'}
         },
         ...mutationOptions
       })`}
    </Function>
  )
}

type FrameworkProps = TemplateProps & {
  context: {
    factory: {
      name: string
    }
  }
}

const defaultTemplates = {
  get react() {
    return function (props: FrameworkProps): ReactNode {
      return <Template {...props} />
    }
  },
  get solid() {
    return function (props: FrameworkProps): ReactNode {
      return <Template {...props} />
    }
  },
  get svelte() {
    return function (props: FrameworkProps): ReactNode {
      return <Template {...props} />
    }
  },
  get vue() {
    return function ({ client, context, ...rest }: FrameworkProps): ReactNode {
      const { factory } = context

      const importNames = getImportNames()

      const hookName = importNames.mutation.vue.hookName
      const resultType = importNames.mutation.vue.resultType
      const optionsType = importNames.mutation.vue.optionsType
      const schemas = useOperationSchemas()
      const params = new FunctionParams()

      const resultGenerics = [`${factory.name}["response"]`, `${factory.name}["error"]`, client.withData ? `${factory.name}["request"]` : 'void', 'unknown']

      params.add([
        ...getASTParams(schemas.pathParams, {
          typed: true,
          override: (item) => ({
            ...item,
            name: item.name ? `ref${transformers.pascalCase(item.name)}` : undefined,
            enabled: !!item.name,
            type: `MaybeRef<${item.type}>`,
          }),
        }),
        {
          name: 'refParams',
          type: `MaybeRef<${schemas.queryParams?.name}>`,
          enabled: client.withQueryParams,
          required: isRequired(schemas.queryParams?.schema),
        },
        {
          name: 'refHeaders',
          type: `MaybeRef<${schemas.headerParams?.name}>`,
          enabled: client.withHeaders,
          required: isRequired(schemas.headerParams?.schema),
        },
        {
          name: 'options',
          type: `{
        mutation?: ${optionsType}<${resultGenerics.join(', ')}>,
        client?: ${factory.name}['client']['parameters']
    }`,
          default: '{}',
        },
      ])

      const unrefs = params.items
        .filter((item) => item.enabled)
        .map((item) => {
          return item.name ? `const ${transformers.camelCase(item.name.replace('ref', ''))} = unref(${item.name})` : undefined
        })
        .join('\n')

      const hook = {
        name: hookName,
        generics: [`${factory.name}['response']`, `${factory.name}["error"]`, client.withData ? `${factory.name}["request"]` : 'void', 'unknown'].join(', '),
        children: unrefs,
      }

      return (
        <>
          <Template
            {...rest}
            mutateParams="data"
            params={params.toString()}
            returnType={`${resultType}<${resultGenerics.join(', ')}>`}
            client={client}
            hook={hook}
          />
        </>
      )
    }
  },
} as const

type Props = {
  factory: {
    name: string
  }
  resultType: string
  hookName: string
  optionsType: string
  /**
   * This will make it possible to override the default behaviour.
   */
  Template?: React.ComponentType<FrameworkProps>
}

export function Mutation({ factory, resultType, hookName, optionsType, Template = defaultTemplates.react }: Props): ReactNode {
  // TODO do checks on pathParamsType
  const {
    options: { dataReturnType, mutate },
  } = usePlugin<PluginOptions>()

  if (!mutate) {
    return null
  }

  const operation = useOperation()
  const name = useOperationName({ type: 'function' })
  const schemas = useOperationSchemas()

  const params = new FunctionParams()
  const mutateParams = new FunctionParams()

  const requestType =
    mutate.variablesType === 'mutate'
      ? FunctionParams.toObject([
          ...getASTParams(schemas.pathParams, { typed: true }),
          {
            name: 'params',
            type: `${factory.name}['queryParams']`,
            enabled: !!schemas.queryParams?.name,
            required: isRequired(schemas.queryParams?.schema),
          },
          {
            name: 'headers',
            type: `${factory.name}['headerParams']`,
            enabled: !!schemas.headerParams?.name,
            required: isRequired(schemas.headerParams?.schema),
          },
          {
            name: 'data',
            type: `${factory.name}['request']`,
            enabled: !!schemas.request?.name,
            required: isRequired(schemas.request?.schema),
          },
        ])?.type
      : schemas.request?.name
        ? `${factory.name}['request']`
        : 'never'

  const client = {
    method: operation.method,
    path: new URLPath(operation.path),
    generics: [`${factory.name}["data"]`, `${factory.name}["error"]`, requestType ? `${factory.name}["request"]` : 'void'].join(', '),
    withQueryParams: !!schemas.queryParams?.name,
    withData: !!schemas.request?.name,
    withPathParams: !!schemas.pathParams?.name,
    withHeaders: !!schemas.headerParams?.name,
  }
  const hook = {
    name: hookName,
    generics: [`${factory.name}['response']`, `${factory.name}["error"]`, requestType ? `${requestType}` : 'void'].join(', '),
  }

  const resultGenerics = [
    `${factory.name}["response"]`,
    `${factory.name}["error"]`,
    mutate?.variablesType === 'mutate' ? requestType : `${factory.name}["request"]`,
  ]

  if (mutate?.variablesType === 'mutate') {
    params.add([
      {
        name: 'options',
        type: `{
      mutation?: ${optionsType}<${resultGenerics.join(', ')}>,
      client?: ${factory.name}['client']['parameters']
  }`,
        default: '{}',
      },
    ])

    mutateParams.add([
      [
        ...getASTParams(schemas.pathParams, { typed: false }),
        {
          name: 'params',
          enabled: client.withQueryParams,
          required: isRequired(schemas.queryParams?.schema),
        },
        {
          name: 'headers',
          enabled: client.withHeaders,
          required: isRequired(schemas.headerParams?.schema),
        },
        {
          name: 'data',
          enabled: !!schemas.request?.name,
          required: isRequired(schemas.request?.schema),
        },
      ],
    ])
  } else {
    params.add([
      ...getASTParams(schemas.pathParams, { typed: true }),
      {
        name: 'params',
        type: `${factory.name}['queryParams']`,
        enabled: client.withQueryParams,
        required: isRequired(schemas.queryParams?.schema),
      },
      {
        name: 'headers',
        type: `${factory.name}['headerParams']`,
        enabled: client.withHeaders,
        required: isRequired(schemas.headerParams?.schema),
      },
      {
        name: 'options',
        type: `{
      mutation?: ${optionsType}<${resultGenerics.join(', ')}>,
      client?: ${factory.name}['client']['parameters']
  }`,
        default: '{}',
      },
    ])

    mutateParams.add([
      {
        name: 'data',
        enabled: !!schemas.request?.name,
        required: isRequired(schemas.request?.schema),
      },
    ])
  }

  return (
    <>
      <Template
        name={name}
        JSDoc={{ comments: getComments(operation) }}
        client={client}
        hook={hook}
        params={params.toString()}
        mutateParams={mutateParams.toString()}
        returnType={`${resultType}<${resultGenerics.join(', ')}>`}
        dataReturnType={dataReturnType}
        context={{ factory }}
      />
    </>
  )
}

type FileProps = {
  /**
   * This will make it possible to override the default behaviour.
   */
  templates?: typeof defaultTemplates
  /**
   * This will make it possible to override the default behaviour.
   */
  imports?: typeof MutationImports.templates
}

Mutation.File = function ({ templates = defaultTemplates, imports = MutationImports.templates }: FileProps): ReactNode {
  const {
    options: {
      client: { importPath },
      framework,
    },
  } = usePlugin<PluginOptions>()
  const schemas = useOperationSchemas()
  const file = useGetOperationFile()
  const fileType = useGetOperationFile({ pluginKey: swaggerTsPluginKey })
  const factoryName = useOperationName({ type: 'type' })

  const importNames = getImportNames()
  const Template = templates[framework]
  const Import = imports[framework]
  const factory = {
    name: factoryName,
  }

  return (
    <Editor language="typescript">
      <File<FileMeta> baseName={file.baseName} path={file.path} meta={file.meta}>
        <File.Import name={'client'} path={importPath} />
        <File.Import name={['ResponseConfig']} path={importPath} isTypeOnly />
        <File.Import
          name={[
            schemas.request?.name,
            schemas.response.name,
            schemas.pathParams?.name,
            schemas.queryParams?.name,
            schemas.headerParams?.name,
            ...(schemas.errors?.map((error) => error.name) || []),
          ].filter(Boolean)}
          root={file.path}
          path={fileType.path}
          isTypeOnly
        />
        <MutationImports Template={Import} />
        <File.Source>
          <SchemaType factory={factory} />
          <Mutation
            factory={factory}
            Template={Template}
            hookName={importNames.mutation[framework].hookName}
            resultType={importNames.mutation[framework].resultType}
            optionsType={importNames.mutation[framework].optionsType}
          />
        </File.Source>
      </File>
    </Editor>
  )
}

Mutation.templates = defaultTemplates
