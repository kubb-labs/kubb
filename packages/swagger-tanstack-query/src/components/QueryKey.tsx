import { FunctionParams, URLPath } from '@kubb/core/utils'
import { Function, Type, usePlugin } from '@kubb/react'
import { useOperation, useOperationSchemas } from '@kubb/swagger/hooks'
import { getASTParams, isRequired } from '@kubb/swagger/utils'

import type { ReactNode } from 'react'
import type { PluginOptions } from '../types'

type TemplateProps = {
  /**
   * Name of the function
   */
  name: string
  /**
   * TypeName of the function in PascalCase
   */
  typeName: string
  /**
   * Parameters/options/props that need to be used
   */
  params: string
  /**
   * Generics that needs to be added for TypeScript
   */
  generics?: string
  /**
   * ReturnType(see async for adding Promise type)
   */
  returnType?: string
  /**
   * Options for JSdocs
   */
  JSDoc?: {
    comments: string[]
  }
  keys?: string
}

function Template({ name, typeName, params, generics, returnType, JSDoc, keys }: TemplateProps): ReactNode {
  return (
    <>
      <Function.Arrow name={name} export generics={generics} params={params} returnType={returnType} singleLine JSDoc={JSDoc}>
        {`[${keys}] as const`}
      </Function.Arrow>

      <Type name={typeName} export>
        {`ReturnType<typeof ${name}>`}
      </Type>
    </>
  )
}

type FrameworkProps = TemplateProps & {
  context: {
    factory: {
      name: string
    }
  }
}

const defaultTemplates = {
  get react() {
    return function (props: FrameworkProps): ReactNode {
      return <Template {...props} />
    }
  },
  get solid() {
    return function (props: FrameworkProps): ReactNode {
      return <Template {...props} />
    }
  },
  get svelte() {
    return function (props: FrameworkProps): ReactNode {
      return <Template {...props} />
    }
  },
  get vue() {
    return function ({ context, ...rest }: FrameworkProps): ReactNode {
      const { factory } = context

      const {
        options: { pathParamsType },
      } = usePlugin<PluginOptions>()
      const schemas = useOperationSchemas()
      const operation = useOperation()
      const path = new URLPath(operation.path)
      const params = new FunctionParams()
      const withQueryParams = !!schemas.queryParams?.name

      params.add([
        ...getASTParams(schemas.pathParams, {
          typed: true,
          asObject: pathParamsType === 'object',
          override: (item) => ({ ...item, type: `MaybeRef<${item.type}>` }),
        }),
        {
          name: 'params',
          type: schemas.queryParams?.name ? `MaybeRef<${`${factory.name}["queryParams"]`}>` : undefined,
          enabled: !!schemas.queryParams?.name,
          required: isRequired(schemas.queryParams?.schema),
        },
      ])

      const keys = [
        path.toObject({
          type: 'path',
          stringify: true,
          replacer: (pathParam) => `unref(${pathParam})`,
        }),
        withQueryParams ? '...(params ? [params] : [])' : undefined,
      ].filter(Boolean)

      return <Template {...rest} params={params.toString()} keys={keys.join(', ')} />
    }
  },
} as const

type Props = {
  name: string
  typeName: string
  keysFn: (keys: unknown[]) => unknown[]
  factory: {
    name: string
  }
  /**
   * This will make it possible to override the default behaviour.
   */
  Template?: React.ComponentType<FrameworkProps>
}

export function QueryKey({ name, typeName, factory, keysFn, Template = defaultTemplates.react }: Props): ReactNode {
  const {
    options: { pathParamsType },
  } = usePlugin<PluginOptions>()
  const schemas = useOperationSchemas()
  const operation = useOperation()
  const path = new URLPath(operation.path)
  const params = new FunctionParams()
  const withQueryParams = !!schemas.queryParams?.name

  params.add([
    ...getASTParams(schemas.pathParams, {
      typed: true,
      asObject: pathParamsType === 'object',
    }),
    {
      name: 'params',
      type: `${factory.name}["queryParams"]`,
      enabled: !!schemas.queryParams?.name,
      required: isRequired(schemas.queryParams?.schema),
    },
  ])

  const keys = [
    path.toObject({
      type: 'path',
      stringify: true,
    }),
    withQueryParams ? '...(params ? [params] : [])' : undefined,
  ].filter(Boolean)

  return <Template typeName={typeName} name={name} params={params.toString()} keys={keysFn(keys).join(', ')} context={{ factory }} />
}

QueryKey.templates = defaultTemplates
