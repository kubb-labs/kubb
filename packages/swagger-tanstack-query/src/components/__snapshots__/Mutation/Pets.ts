type CreatePetsClient = typeof client<CreatePetsMutationResponse, never, CreatePetsMutationRequest>
type CreatePets = {
  data: CreatePetsMutationResponse
  error: never
  request: CreatePetsMutationRequest
  pathParams: never
  queryParams: never
  headerParams: never
  response: CreatePetsMutationResponse
  client: {
    parameters: Partial<Parameters<CreatePetsClient>[0]>
    return: Awaited<ReturnType<CreatePetsClient>>
  }
}

export const CreatePetsQueryKey = () => [{ url: '/pets' }] as const
export type CreatePetsQueryKey = ReturnType<typeof CreatePetsQueryKey>
export function CreatePetsQueryOptions<TData = CreatePets['response'], TQueryData = CreatePets['response']>(
  options: CreatePets['client']['parameters'] = {},
): WithRequired<UseBaseQueryOptions<CreatePets['response'], CreatePets['error'], TData, TQueryData>, 'queryKey'> {
  const queryKey = CreatePetsQueryKey()

  return {
    queryKey,
    queryFn: async () => {
      const res = await client<CreatePets['data'], CreatePets['error']>({
        method: 'post',
        url: `/pets`,
        data,
        ...options,
      })

      return res.data
    },
  }
}
/**
 * @summary Create a pet
 * @link /pets */

export function useCreatePets<TData = CreatePets['response'], TQueryData = CreatePets['response'], TQueryKey extends QueryKey = CreatePetsQueryKey>(
  options: {
    query?: Partial<UseBaseQueryOptions<CreatePets['response'], CreatePets['error'], TData, TQueryData, TQueryKey>>
    client?: CreatePets['client']['parameters']
  } = {},
): UseQueryResult<TData, CreatePets['error']> & { queryKey: TQueryKey } {
  const { query: queryOptions, client: clientOptions = {} } = options ?? {}
  const queryKey = queryOptions?.queryKey ?? CreatePetsQueryKey()

  const query = useQuery<CreatePets['data'], CreatePets['error'], TData, any>({
    ...CreatePetsQueryOptions<TData, TQueryData>(clientOptions),
    queryKey,
    ...queryOptions,
  }) as UseQueryResult<TData, CreatePets['error']> & { queryKey: TQueryKey }

  query.queryKey = queryKey as TQueryKey

  return query
}
