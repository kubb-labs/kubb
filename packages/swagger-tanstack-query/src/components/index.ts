export { Mutation } from './Mutation.tsx'
export { Operations } from './Operations.tsx'
export { Query } from './Query.tsx'
export { QueryKey } from './QueryKey.tsx'
export { QueryOptions } from './QueryOptions.tsx'
