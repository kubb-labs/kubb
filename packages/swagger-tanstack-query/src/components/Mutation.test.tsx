import { mockedPluginManager } from '@kubb/core/mocks'
import { camelCase, pascalCase } from '@kubb/core/transformers'
import { createRootServer } from '@kubb/react/server'
import { OasManager } from '@kubb/swagger'
import { Oas } from '@kubb/swagger/components'

import { OperationGenerator } from '../OperationGenerator.tsx'
import { Mutation } from './Mutation.tsx'
import { Query } from './Query.tsx'
import { QueryKey } from './QueryKey.tsx'
import { QueryOptions } from './QueryOptions.tsx'

import type { Plugin, ResolveNameParams } from '@kubb/core'
import type { AppContextProps } from '@kubb/react'
import type { GetOperationGeneratorOptions } from '@kubb/swagger'
import type { PluginOptions } from '../types.ts'

describe('<Mutation/>', async () => {
  const oas = await OasManager.parseFromConfig({
    root: './',
    output: { path: 'test', clean: true },
    input: { path: 'packages/swagger-tanstack-query/mocks/petStore.yaml' },
  })
  mockedPluginManager.resolveName = ({ type, name }: ResolveNameParams) => {
    if (type === 'file' || type === 'function') {
      return camelCase(`use ${name}`)
    }

    if (type === 'type') {
      return pascalCase(name)
    }
    return name
  }

  test('pets with veriableType `hook`', async () => {
    const options: GetOperationGeneratorOptions<OperationGenerator> = {
      framework: 'react',
      infinite: false,
      suspense: false,
      dataReturnType: 'data',
      pathParamsType: 'inline',
      templates: {
        query: Query.templates,
        queryKey: QueryKey.templates,
        queryOptions: QueryOptions.templates,
        mutation: Mutation.templates,
      },
      client: {
        importPath: '@kubb/swagger-client/client',
      },
      parser: undefined,
      query: false,
      queryOptions: false,
      mutate: { variablesType: 'hook', methods: ['post'] },
    }

    const plugin = { options } as Plugin<PluginOptions>
    const og = await new OperationGenerator(options, {
      oas,
      exclude: [],
      include: undefined,
      pluginManager: mockedPluginManager,
      plugin,
      contentType: undefined,
      override: undefined,
    })

    const operation = oas.operation('/pets/{uuid}', 'post')
    const context: AppContextProps<PluginOptions['appMeta']> = {
      meta: { pluginManager: mockedPluginManager, plugin },
    }

    const Component = () => {
      return (
        <Oas oas={oas} operations={[operation]} getOperationSchemas={(...props) => og.getSchemas(...props)}>
          <Oas.Operation operation={operation}>
            <Mutation.File />
          </Oas.Operation>
        </Oas>
      )
    }
    const root = createRootServer({ logger: mockedPluginManager.logger })
    const output = await root.renderToString(<Component />, context)

    expect(output).toMatchFileSnapshot('./__snapshots__/gen/useCreatePets.ts')
  })

  test('pets with veriableType `mutate`', async () => {
    const options: GetOperationGeneratorOptions<OperationGenerator> = {
      framework: 'react',
      infinite: false,
      suspense: false,
      dataReturnType: 'data',
      pathParamsType: 'inline',
      templates: {
        query: Query.templates,
        queryKey: QueryKey.templates,
        queryOptions: QueryOptions.templates,
        mutation: Mutation.templates,
      },
      client: {
        importPath: '@kubb/swagger-client/client',
      },
      parser: undefined,
      query: false,
      queryOptions: false,
      mutate: { variablesType: 'mutate', methods: ['post'] },
    }

    const plugin = { options } as Plugin<PluginOptions>
    const og = await new OperationGenerator(options, {
      oas,
      exclude: [],
      include: undefined,
      pluginManager: mockedPluginManager,
      plugin,
      contentType: undefined,
      override: undefined,
    })

    const operation = oas.operation('/pets/{uuid}', 'post')
    const context: AppContextProps<PluginOptions['appMeta']> = {
      meta: { pluginManager: mockedPluginManager, plugin },
    }

    const Component = () => {
      return (
        <Oas oas={oas} operations={[operation]} getOperationSchemas={(...props) => og.getSchemas(...props)}>
          <Oas.Operation operation={operation}>
            <Mutation.File />
          </Oas.Operation>
        </Oas>
      )
    }
    const root = createRootServer({ logger: mockedPluginManager.logger })
    const output = await root.renderToString(<Component />, context)

    expect(output).toMatchFileSnapshot('./__snapshots__/gen/useCreatePetsMutate.ts')
  })
})
