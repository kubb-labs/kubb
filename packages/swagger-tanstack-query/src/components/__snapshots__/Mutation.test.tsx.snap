// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Mutation/> > pets 1`] = `""`;
